File: ccx/parser.py
"""Parse Claude Code session JSONL files and extract the post-compaction slice."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Iterable, Optional


@dataclass
class Turn:
    role: str
    text: str
    tool_calls: list[str] = field(default_factory=list)
    timestamp: str = ""


@dataclass
class Session:
    project_path: str
    session_id: str
    git_branch: Optional[str]
    summary: Optional[str]
    compact_meta: Optional[dict]
    compact_timestamp: Optional[str]
    turns: list[Turn]


def load_jsonl(path: str | Path) -> list[dict]:
    events: list[dict] = []
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                events.append(json.loads(line))
            except json.JSONDecodeError:
                continue
    return events


def extract_text_from_message(message: dict) -> str:
    """Get plain text from a message.content (string or list-of-parts)."""
    content = message.get("content", "")
    if isinstance(content, str):
        return content
    if not isinstance(content, list):
        return ""
    parts: list[str] = []
    for part in content:
        if not isinstance(part, dict):
            continue
        if part.get("type") != "text":
            continue
        text = part.get("text", "") or ""
        if text.startswith("<ide_opened_file>") or text.startswith("<ide_selection>"):
            continue
        parts.append(text)
    return "\n".join(parts)


def find_last_compaction(events: list[dict]) -> tuple[Optional[int], Optional[str], Optional[dict]]:
    """Return (idx, summary_text, compact_metadata) for the last isCompactSummary message,
    or (None, None, None) if the session has never been compacted."""
    summary_idx: Optional[int] = None
    summary_text: Optional[str] = None
    for i in range(len(events) - 1, -1, -1):
        ev = events[i]
        if ev.get("type") == "user" and ev.get("isCompactSummary"):
            summary_idx = i
            summary_text = extract_text_from_message(ev.get("message", {}))
            break

    if summary_idx is None:
        return None, None, None

    compact_meta: Optional[dict] = None
    lower = max(-1, summary_idx - 5)
    for i in range(summary_idx, lower, -1):
        ev = events[i]
        if ev.get("type") == "system" and "compactMetadata" in ev:
            compact_meta = ev.get("compactMetadata")
            break

    return summary_idx, summary_text, compact_meta


def summarize_tool_use(part: dict) -> str:
    name = part.get("name", "?")
    inp = part.get("input") or {}
    if not isinstance(inp, dict):
        inp = {}

    if name == "Bash":
        desc = inp.get("description") or ""
        cmd = inp.get("command") or ""
        if desc:
            return f"↳ Bash: {desc}"
        return f"↳ Bash: {cmd[:80]}"
    if name in ("Read", "Write"):
        return f"↳ {name}: {inp.get('file_path', '?')}"
    if name == "Edit":
        return f"↳ Edit: {inp.get('file_path', '?')}"
    if name == "NotebookEdit":
        return f"↳ NotebookEdit: {inp.get('notebook_path', '?')}"
    if name == "Grep":
        return f"↳ Grep: {inp.get('pattern', '?')}"
    if name == "Glob":
        return f"↳ Glob: {inp.get('pattern', '?')}"
    if name == "WebFetch":
        return f"↳ WebFetch: {inp.get('url', '?')}"
    if name == "WebSearch":
        return f"↳ WebSearch: {inp.get('query', '?')}"
    if name == "TodoWrite":
        n = len(inp.get("todos", []) or [])
        return f"↳ TodoWrite: {n} items"
    if name in ("Task", "Agent"):
        desc = inp.get("description") or ""
        sub = inp.get("subagent_type")
        label = f"{sub}: " if sub else ""
        return f"↳ Agent: {label}{desc}"
    return f"↳ {name}"


def event_to_turn(event: dict, include_tool_calls: bool = True) -> Optional[Turn]:
    et = event.get("type")
    if et not in ("user", "assistant"):
        return None
    if event.get("isSidechain"):
        return None
    if event.get("isCompactSummary"):
        return None

    msg = event.get("message", {}) or {}
    content = msg.get("content", [])

    if et == "user":
        if isinstance(content, list):
            has_real_text = False
            for part in content:
                if not isinstance(part, dict):
                    continue
                if part.get("type") == "text":
                    text = part.get("text", "") or ""
                    if text and not (
                        text.startswith("<ide_opened_file>")
                        or text.startswith("<ide_selection>")
                    ):
                        has_real_text = True
                        break
            if not has_real_text:
                return None
        text = extract_text_from_message(msg)
        if not text.strip():
            return None
        return Turn(role="user", text=text, timestamp=event.get("timestamp", ""))

    text_parts: list[str] = []
    tool_calls: list[str] = []
    if isinstance(content, list):
        for part in content:
            if not isinstance(part, dict):
                continue
            t = part.get("type")
            if t == "text":
                txt = part.get("text", "") or ""
                if txt.strip():
                    text_parts.append(txt)
            elif t == "tool_use" and include_tool_calls:
                tool_calls.append(summarize_tool_use(part))
    elif isinstance(content, str):
        if content.strip():
            text_parts.append(content)

    text = "\n".join(text_parts).strip()
    if not text and not tool_calls:
        return None

    return Turn(
        role="assistant",
        text=text,
        tool_calls=tool_calls,
        timestamp=event.get("timestamp", ""),
    )


def _first_meta(events: Iterable[dict]) -> dict[str, Any]:
    for ev in events:
        if "cwd" in ev:
            return {
                "cwd": ev.get("cwd", ""),
                "sessionId": ev.get("sessionId", ""),
                "gitBranch": ev.get("gitBranch", ""),
            }
    return {"cwd": "", "sessionId": "", "gitBranch": ""}


def parse_session(
    jsonl_path: str | Path,
    include_tool_calls: bool = True,
    max_turns: Optional[int] = None,
) -> Session:
    jsonl_path = Path(jsonl_path)
    events = load_jsonl(jsonl_path)
    meta = _first_meta(events)

    summary_idx, summary_text, compact_meta = find_last_compaction(events)
    compact_timestamp: Optional[str] = None
    if summary_idx is not None:
        compact_timestamp = events[summary_idx].get("timestamp")
        post_events = events[summary_idx + 1 :]
    else:
        post_events = events

    turns: list[Turn] = []
    for ev in post_events:
        turn = event_to_turn(ev, include_tool_calls=include_tool_calls)
        if turn is not None:
            turns.append(turn)

    if max_turns is not None and max_turns > 0:
        turns = turns[-max_turns:]

    return Session(
        project_path=meta["cwd"],
        session_id=meta["sessionId"] or jsonl_path.stem,
        git_branch=meta["gitBranch"] or None,
        summary=summary_text,
        compact_meta=compact_meta,
        compact_timestamp=compact_timestamp,
        turns=turns,
    )

File: internal/parser/parser.go
// Package parser loads Claude Code session JSONL files and extracts the
// post-compaction conversation slice.
package parser

import (
	"bufio"
	"encoding/json"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// Turn is a single user or assistant turn in a session.
type Turn struct {
	Role      string   `json:"role"`
	Text      string   `json:"text"`
	ToolCalls []string `json:"tool_calls"`
	Timestamp string   `json:"timestamp"`
	UUID      string   `json:"uuid,omitempty"`
}

// CompactMeta holds the raw compactMetadata fields from a system event.
type CompactMeta struct {
	Trigger    string `json:"trigger,omitempty"`
	PreTokens  int    `json:"preTokens,omitempty"`
	PostTokens int    `json:"postTokens,omitempty"`
	DurationMs int    `json:"durationMs,omitempty"`
	// Extra holds fields not modeled above so render can pass them through.
	Extra map[string]any `json:"-"`
}

// Session is the parsed view of a session JSONL.
type Session struct {
	ProjectPath      string
	SessionID        string
	GitBranch        string
	Summary          string
	HasSummary       bool
	CompactMeta      *CompactMeta
	CompactTimestamp string
	Turns            []Turn

	// Events holds the full parsed event stream for callers (state/cursor) that
	// need to slice or look up by uuid.
	Events []Event
	// SummaryIdx is the index of the last isCompactSummary event in Events,
	// or -1 if the session has never been compacted.
	SummaryIdx int
}

// Event is a single JSONL line as a parsed object plus the raw 0-based line
// index it came from (useful as a fallback cursor).
type Event struct {
	Raw       map[string]any
	LineIndex int
}

// LoadJSONL reads a JSONL file, skipping blank lines and silently dropping
// malformed JSON lines.
func LoadJSONL(path string) ([]Event, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	return loadFromReader(f)
}

func loadFromReader(r io.Reader) ([]Event, error) {
	var events []Event
	sc := bufio.NewScanner(r)
	sc.Buffer(make([]byte, 0, 1<<16), 1<<25)
	idx := 0
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			idx++
			continue
		}
		var obj map[string]any
		if err := json.Unmarshal([]byte(line), &obj); err != nil {
			idx++
			continue
		}
		events = append(events, Event{Raw: obj, LineIndex: idx})
		idx++
	}
	if err := sc.Err(); err != nil {
		return events, err
	}
	return events, nil
}

// ExtractTextFromMessage returns plain text from a message.content (string or
// list-of-parts), skipping IDE markers and non-text parts.
func ExtractTextFromMessage(message map[string]any) string {
	content, ok := message["content"]
	if !ok {
		return ""
	}
	if s, ok := content.(string); ok {
		return s
	}
	parts, ok := content.([]any)
	if !ok {
		return ""
	}
	var out []string
	for _, p := range parts {
		part, ok := p.(map[string]any)
		if !ok {
			continue
		}
		if part["type"] != "text" {
			continue
		}
		text, _ := part["text"].(string)
		if strings.HasPrefix(text, "<ide_opened_file>") || strings.HasPrefix(text, "<ide_selection>") {
			continue
		}
		out = append(out, text)
	}
	return strings.Join(out, "\n")
}

// FindLastCompaction scans events backwards for the last isCompactSummary user
// event and looks back ≤5 events for the matching system+compactMetadata.
// Returns (idx, summaryText, meta) or (-1, "", nil) if never compacted.
func FindLastCompaction(events []Event) (int, string, *CompactMeta) {
	summaryIdx := -1
	var summaryText string
	for i := len(events) - 1; i >= 0; i-- {
		ev := events[i].Raw
		if ev["type"] == "user" && truthy(ev["isCompactSummary"]) {
			summaryIdx = i
			msg, _ := ev["message"].(map[string]any)
			if msg != nil {
				summaryText = ExtractTextFromMessage(msg)
			}
			break
		}
	}
	if summaryIdx < 0 {
		return -1, "", nil
	}

	lower := summaryIdx - 5
	if lower < -1 {
		lower = -1
	}
	for i := summaryIdx; i > lower; i-- {
		ev := events[i].Raw
		if ev["type"] != "system" {
			continue
		}
		meta, ok := ev["compactMetadata"].(map[string]any)
		if !ok {
			continue
		}
		return summaryIdx, summaryText, mapToCompactMeta(meta)
	}
	return summaryIdx, summaryText, nil
}

func mapToCompactMeta(m map[string]any) *CompactMeta {
	cm := &CompactMeta{Extra: map[string]any{}}
	for k, v := range m {
		switch k {
		case "trigger":
			if s, ok := v.(string); ok {
				cm.Trigger = s
			}
		case "preTokens":
			cm.PreTokens = toInt(v)
		case "postTokens":
			cm.PostTokens = toInt(v)
		case "durationMs":
			cm.DurationMs = toInt(v)
		default:
			cm.Extra[k] = v
		}
	}
	return cm
}

func toInt(v any) int {
	switch x := v.(type) {
	case float64:
		return int(x)
	case int:
		return x
	case int64:
		return int(x)
	}
	return 0
}

// SummarizeToolUse renders a single tool_use part into the `↳ Tool: ...` line
// shown in the markdown output. Format must match the Python implementation
// exactly — tests pin it.
func SummarizeToolUse(part map[string]any) string {
	name, _ := part["name"].(string)
	if name == "" {
		name = "?"
	}
	inp, _ := part["input"].(map[string]any)
	if inp == nil {
		inp = map[string]any{}
	}
	getStr := func(k string) string {
		s, _ := inp[k].(string)
		return s
	}

	switch name {
	case "Bash":
		desc := getStr("description")
		cmd := getStr("command")
		if desc != "" {
			return "↳ Bash: " + desc
		}
		if len(cmd) > 80 {
			cmd = cmd[:80]
		}
		return "↳ Bash: " + cmd
	case "Read", "Write":
		fp := getStr("file_path")
		if fp == "" {
			fp = "?"
		}
		return "↳ " + name + ": " + fp
	case "Edit":
		fp := getStr("file_path")
		if fp == "" {
			fp = "?"
		}
		return "↳ Edit: " + fp
	case "NotebookEdit":
		nb := getStr("notebook_path")
		if nb == "" {
			nb = "?"
		}
		return "↳ NotebookEdit: " + nb
	case "Grep":
		p := getStr("pattern")
		if p == "" {
			p = "?"
		}
		return "↳ Grep: " + p
	case "Glob":
		p := getStr("pattern")
		if p == "" {
			p = "?"
		}
		return "↳ Glob: " + p
	case "WebFetch":
		u := getStr("url")
		if u == "" {
			u = "?"
		}
		return "↳ WebFetch: " + u
	case "WebSearch":
		q := getStr("query")
		if q == "" {
			q = "?"
		}
		return "↳ WebSearch: " + q
	case "TodoWrite":
		todos, _ := inp["todos"].([]any)
		return "↳ TodoWrite: " + strconv.Itoa(len(todos)) + " items"
	case "Task", "Agent":
		desc := getStr("description")
		sub := getStr("subagent_type")
		label := ""
		if sub != "" {
			label = sub + ": "
		}
		return "↳ Agent: " + label + desc
	}
	return "↳ " + name
}

// EventToTurn converts a single event into a Turn, applying the noise filter.
// Returns nil if the event should be dropped from the turn stream.
func EventToTurn(event Event, includeToolCalls bool) *Turn {
	ev := event.Raw
	et, _ := ev["type"].(string)
	if et != "user" && et != "assistant" {
		return nil
	}
	if truthy(ev["isSidechain"]) {
		return nil
	}
	if truthy(ev["isCompactSummary"]) {
		return nil
	}

	msg, _ := ev["message"].(map[string]any)
	if msg == nil {
		msg = map[string]any{}
	}
	content := msg["content"]
	ts, _ := ev["timestamp"].(string)
	uuid, _ := ev["uuid"].(string)

	if et == "user" {
		if parts, ok := content.([]any); ok {
			hasReal := false
			for _, p := range parts {
				part, ok := p.(map[string]any)
				if !ok {
					continue
				}
				if part["type"] != "text" {
					continue
				}
				text, _ := part["text"].(string)
				if text == "" {
					continue
				}
				if strings.HasPrefix(text, "<ide_opened_file>") || strings.HasPrefix(text, "<ide_selection>") {
					continue
				}
				hasReal = true
				break
			}
			if !hasReal {
				return nil
			}
		}
		text := ExtractTextFromMessage(msg)
		if strings.TrimSpace(text) == "" {
			return nil
		}
		return &Turn{Role: "user", Text: text, Timestamp: ts, UUID: uuid}
	}

	// assistant
	var textParts []string
	var toolCalls []string
	switch c := content.(type) {
	case []any:
		for _, p := range c {
			part, ok := p.(map[string]any)
			if !ok {
				continue
			}
			t, _ := part["type"].(string)
			switch t {
			case "text":
				txt, _ := part["text"].(string)
				if strings.TrimSpace(txt) != "" {
					textParts = append(textParts, txt)
				}
			case "tool_use":
				if includeToolCalls {
					toolCalls = append(toolCalls, SummarizeToolUse(part))
				}
			}
		}
	case string:
		if strings.TrimSpace(c) != "" {
			textParts = append(textParts, c)
		}
	}

	text := strings.TrimSpace(strings.Join(textParts, "\n"))
	if text == "" && len(toolCalls) == 0 {
		return nil
	}
	return &Turn{Role: "assistant", Text: text, ToolCalls: toolCalls, Timestamp: ts, UUID: uuid}
}

func truthy(v any) bool {
	b, ok := v.(bool)
	return ok && b
}

// ParseSession loads and parses a session JSONL into a Session.
//
// includeToolCalls controls whether assistant tool_use parts are summarized;
// maxTurns (when > 0) keeps only the last N turns.
func ParseSession(jsonlPath string, includeToolCalls bool, maxTurns int) (*Session, error) {
	events, err := LoadJSONL(jsonlPath)
	if err != nil {
		return nil, err
	}
	return parseEvents(events, jsonlPath, includeToolCalls, maxTurns), nil
}

func parseEvents(events []Event, jsonlPath string, includeToolCalls bool, maxTurns int) *Session {
	cwd, sessionID, branch := firstMeta(events)
	if sessionID == "" {
		sessionID = stem(jsonlPath)
	}

	summaryIdx, summaryText, compactMeta := FindLastCompaction(events)
	var compactTS string
	postEvents := events
	if summaryIdx >= 0 {
		if ts, ok := events[summaryIdx].Raw["timestamp"].(string); ok {
			compactTS = ts
		}
		postEvents = events[summaryIdx+1:]
	}

	var turns []Turn
	for _, ev := range postEvents {
		if t := EventToTurn(ev, includeToolCalls); t != nil {
			turns = append(turns, *t)
		}
	}

	if maxTurns > 0 && len(turns) > maxTurns {
		turns = turns[len(turns)-maxTurns:]
	}

	return &Session{
		ProjectPath:      cwd,
		SessionID:        sessionID,
		GitBranch:        branch,
		Summary:          summaryText,
		HasSummary:       summaryIdx >= 0,
		CompactMeta:      compactMeta,
		CompactTimestamp: compactTS,
		Turns:            turns,
		Events:           events,
		SummaryIdx:       summaryIdx,
	}
}

// ParseSessionFromEvents parses an already-loaded event slice. Used by callers
// that have read events for cursor logic and want to avoid re-reading the file.
func ParseSessionFromEvents(events []Event, jsonlPath string, includeToolCalls bool, maxTurns int) *Session {
	return parseEvents(events, jsonlPath, includeToolCalls, maxTurns)
}

// TurnsFromEvents converts a slice of events into turns with the same noise
// filter, useful for diff slices that aren't a full session parse.
func TurnsFromEvents(events []Event, includeToolCalls bool) []Turn {
	var turns []Turn
	for _, ev := range events {
		if t := EventToTurn(ev, includeToolCalls); t != nil {
			turns = append(turns, *t)
		}
	}
	return turns
}

func firstMeta(events []Event) (cwd, sessionID, branch string) {
	for _, ev := range events {
		if _, ok := ev.Raw["cwd"]; !ok {
			continue
		}
		cwd, _ = ev.Raw["cwd"].(string)
		sessionID, _ = ev.Raw["sessionId"].(string)
		branch, _ = ev.Raw["gitBranch"].(string)
		return
	}
	return "", "", ""
}

func stem(path string) string {
	name := filepath.Base(path)
	if dot := strings.LastIndex(name, "."); dot > 0 {
		return name[:dot]
	}
	return name
}
